gadgets/trace_malloc/program.bpf.c

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2024 The Inspektor Gadget authors */
+
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <gadget/buffer.h>
+#include <gadget/macros.h>
+
+struct event {
+	__u32 pid;
+	__u64 addr;
+	bool is_malloc;
+};
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+
+GADGET_TRACER(open, events, event);
+
+SEC("uretprobe//usr/lib/libc.so.6:malloc")
+int trace_uprobe_malloc(struct pt_regs *ctx)
```

Comment:
On Fedora, I had to use `/lib64/libc.so.6` instead of `/usr/lib/libc.so.6` to make it work. Do you know the difference between those two paths?

Reply:
Replying to myself: x86_64 vs i686.

```
$ file /lib64/libc.so.6
/lib64/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=f888be5f5e7d58e04cabb8c675c7ab94e77dd68c, for GNU/Linux 3.2.0, not stripped

$ rpm -qf /lib64/libc.so.6
glibc-2.37-16.fc38.x86_64
```

```
$ file /usr/lib/libc.so.6
/usr/lib/libc.so.6: ELF 32-bit LSB shared object, Intel 80386, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux.so.2, BuildID[sha1]=a116613533d396e2c906376b67b7d60b398fe10a, for GNU/Linux 3.2.0, not stripped
$ rpm -qf /usr/lib/libc.so.6
glibc-2.37-16.fc38.i686
```

I guess your distribution uses different paths.

Since different containers use different base Linux distributions, I wonder if we will need to search for libraries in different paths following the same algorithm as defined in [ld.so(8)](https://man7.org/linux/man-pages/man8/ld.so.8.html).

See also [prior art from bpftrace](https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#probes-uprobe):

> When tracing libraries, it is sufficient to specify the library name instead of a full path. The path will be then automatically resolved using /etc/ld.so.cache:
> 
> uprobe:libc:malloc { printf("Allocated %d bytes\n", arg0); }

In any cases, that's something for a future PR.

Reply:
On my computer, the above two files share the same inode. But on ubuntu, it's `/usr/lib32/libc.so.6` and `/usr/lib/x86_64-linux-gnu/libc.so.6`. So it may depends on the default `ld` searching path of the distros.